### PR TITLE
AArch64 macOS: Handling 8/16-bit integer arguments for JNI

### PR DESCRIPTION
**Pull request by knn-k**
This commit changes the layout of JNI arguments that are passed in
the stack.  AArch64 macOS packs 8/16-bit integer arguments to 1/2-byte
boundaries of address space.

Signed-off-by: KONNO Kazuhiro <konno@jp.ibm.com>

### CODE REVIEW

**Comment by knn-k** [comment]
jenkins test sanity alinux64 jdk17

**Comment by knn-k** [comment]
jenkins test sanity amac jdk17

**Comment by knn-k** [comment]
Previous change in AArch64 JNI linkage for macOS: #14194

**Comment by knn-k** [comment]
There are 4 failures in https://openj9-jenkins.osuosl.org/job/Test_openjdk17_j9_sanity.functional_aarch64_mac_Personal/8/tapResults/:
- 1 failure with CryptoTest: It is resolved by #14444
- 3 failures with jniargtests: They are independent of this PR.  They fail even in the interpreter mode, and I am investigating now.

**Comment by knn-k** [comment]
The PR above is ready for review now.  It does not break AArch64 Linux.

**Comment by knn-k** [comment]
jenkins test sanity amac jdk18

**Comment by knn-k** [comment]
Issue for the failures with jniargtests: #14500

**Comment by knn-k** [comment]
@0xdaryl Could you review this PR, please?  Thanks.

File: runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp [hunk 1]
```diff
@@ -542,25 +581,45 @@ size_t J9::ARM64::JNILinkage::buildJNIArgs(TR::Node *callNode, TR::RegisterDepen
...
+                     {
+                     op = TR::InstOpCode::strimmw;
+                     offsetInc = 4;
+                     argOffset = (argOffset + 3) & ~3; // adjust to 4-byte boundary
+                     }
+                  else
+                     {
+                     TR_ASSERT(0, "Unexpected type %c", *argType);
+                     }
+                  }
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp:619]: Same comment as above on simplifying this if/else ladder.  The one wrinkle in this code is that you do need to set `op` to a particular store instruction, but I think a small if/else or switch could accomplish that.

File: runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp [hunk 2]
```diff
@@ -424,7 +437,25 @@ size_t J9::ARM64::JNILinkage::buildJNIArgs(TR::Node *callNode, TR::RegisterDepen
...
+                     }
+                  else if (nativeType == TR::Int32)
+                     {
+                     totalSize = (totalSize + 3) & ~3; // adjust to 4-byte boundary
+                     totalSize += 4;
+                     }
+                  else
+                     {
+                     TR_ASSERT(0, "Unexpected type %c", *argType);
+                     }
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp:458]: Yes, that should work.  I will try.

**Comment by knn-k** [comment]
jenkins test sanity amac jdk17

File: runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp]: I updated the code as suggested.

File: runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp (hunk 1 above)
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp:590]: I updated the code in the same way as the first loop plus a `switch` statement.

File: runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp [hunk 3]
```diff
@@ -424,7 +437,25 @@ size_t J9::ARM64::JNILinkage::buildJNIArgs(TR::Node *callNode, TR::RegisterDepen
...
+                     {
+                     totalSize = (totalSize + 1) & ~1; // adjust to 2-byte boundary
+                     totalSize += 2;
+                     }
+                  else if (nativeType == TR::Int32)
+                     {
+                     totalSize = (totalSize + 3) & ~3; // adjust to 4-byte boundary
+                     totalSize += 4;
+                     }
+                  else
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp]: No `switch` statement here now.

File: runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp [hunk 4]
```diff
@@ -542,25 +581,45 @@ size_t J9::ARM64::JNILinkage::buildJNIArgs(TR::Node *callNode, TR::RegisterDepen
...
+                     }
+                  else if (nativeType == TR::Int32)
+                     {
+                     op = TR::InstOpCode::strimmw;
+                     offsetInc = 4;
+                     argOffset = (argOffset + 3) & ~3; // adjust to 4-byte boundary
+                     }
+                  else
+                     {
+                     TR_ASSERT(0, "Unexpected type %c", *argType);
```
[the remark this answers is not in the document]
Comment (knn-k) [reply runtime/compiler/aarch64/codegen/ARM64JNILinkage.cpp]: No `TR_ASSERT` for unexpected type here now.

**Comment by 0xdaryl** [comment]
We only have an issue open for the jniargtests_2 failure (#14500).  Are the other jniargtests failures seen in the CI testing of this PR known, or are expected to be dups of 14500?

**Comment by knn-k** [comment]
I assume the remaining failures with jniargtests_* are dups of 14500.

**Comment by 0xdaryl** [comment]
Sorry for the delay.  I needed to double check something with the loop over the argument children to confirm it is only evaluating children that are part of the original method signature.  I think it is fine.  Merging.

**Comment by knn-k** [comment]
Failures with `nativeZIZIZIZIZIrZ` seem to have been caused by wrong type declarations in the test code.  PR #14560 will fix it.

**Pull request merged**